File: web/src/api/workflow.ts
import { api } from './client'

export type WorkflowStatus =
  | 'PENDING'
  | 'APPROVED'
  | 'REJECTED'
  | 'RETURNED'
  | 'CANCELLED'
  | 'AUTO_APPROVED'

export type WorkflowActionType =
  | 'START'
  | 'APPROVE'
  | 'REJECT'
  | 'RETURN'
  | 'COMMENT'
  | 'CANCEL'
  | 'AUTO_APPROVE'
  | 'DELEGATE'
  | 'ATTACH_DOCUMENT'
  | 'RESUBMIT'

export interface WorkflowInstance {
  id: string
  definitionId: string
  definitionCode: string
  subjectModule: string
  subjectEntity: string
  subjectId: string
  title: string
  currentStepIndex: number
  currentStepRole?: string | null
  status: WorkflowStatus
  initiatedBy: string
  initiatedAt: string
  completedAt?: string | null
  payload?: string | null
  delegatedTo?: string | null
  /** M174 — how many times this instance has been returned and re-submitted */
  resubmitCount?: number
}

export interface WorkflowAction {
  id: string
  instanceId: string
  stepIndex: number
  stepName?: string | null
  action: WorkflowActionType
  actor: string
  comment?: string | null
  ipAddress?: string | null
  /** M166 — non-null for ATTACH_DOCUMENT actions */
  documentRef?: string | null
  createdAt: string
}

export interface WorkflowDefinitionStep {
  order: number
  name: string
  approverRole: string
  slaHours?: number | null
}

export interface WorkflowDefinition {
  id: string
  code: string
  name: string
  description?: string | null
  autoApprove: boolean
  active: boolean
  steps: WorkflowDefinitionStep[]
}

export const workflowApi = {
  definitions: () =>
    api.get<WorkflowDefinition[]>('/workflow/definitions').then((r) => r.data),
  inbox: () => api.get<WorkflowInstance[]>('/workflow/inbox').then((r) => r.data),
  initiated: () => api.get<WorkflowInstance[]>('/workflow/initiated').then((r) => r.data),
  get: (id: string) =>
    api.get<WorkflowInstance>(`/workflow/instances/${id}`).then((r) => r.data),
  history: (id: string) =>
    api.get<WorkflowAction[]>(`/workflow/instances/${id}/history`).then((r) => r.data),
  bySubject: (module: string, entity: string, id: string) =>
    api
      .get<WorkflowInstance[]>('/workflow/instances', {
        params: { module, entity, id },
      })
      .then((r) => r.data),
  act: (id: string, action: WorkflowActionType, comment?: string, delegateTo?: string, documentRef?: string) =>
    api
      .post<WorkflowInstance>(`/workflow/instances/${id}/actions`, { action, comment, delegateTo, documentRef })
      .then((r) => r.data),
  /** M174 — re-submit a RETURNED instance after corrections */
  resubmit: (id: string, comment?: string) =>
    api
      .post<WorkflowInstance>(`/workflow/instances/${id}/resubmit`, null, {
        params: comment ? { comment } : {},
      })
      .then((r) => r.data),
}

File: .claude/worktrees/agent-abd361ef83a9de69c/mobile/lib/models/workflow.dart
/// Workflow instance returned by GET /api/workflow/instances
class WorkflowInstance {
  final String id;
  final String definitionName;
  final String subjectEntity;
  final String subjectId;
  final String status;
  final String currentStepName;
  final String currentStepRole;
  final String initiatedBy;
  final String initiatedAt;
  final String? completedAt;

  const WorkflowInstance({
    required this.id,
    required this.definitionName,
    required this.subjectEntity,
    required this.subjectId,
    required this.status,
    required this.currentStepName,
    required this.currentStepRole,
    required this.initiatedBy,
    required this.initiatedAt,
    this.completedAt,
  });

  bool get isPending => status == 'PENDING';

  factory WorkflowInstance.fromJson(Map<String, dynamic> j) =>
      WorkflowInstance(
        id: j['id'] as String,
        definitionName: j['definitionName'] as String? ?? '',
        subjectEntity: j['subjectEntity'] as String? ?? '',
        subjectId: j['subjectId'] as String? ?? '',
        status: j['status'] as String? ?? '',
        currentStepName: j['currentStepName'] as String? ?? '',
        currentStepRole: j['currentStepRole'] as String? ?? '',
        initiatedBy: j['initiatedBy'] as String? ?? '',
        initiatedAt: j['initiatedAt'] as String? ?? '',
        completedAt: j['completedAt'] as String?,
      );
}

/// Payslip summary from GET /api/self/payslips
class Payslip {
  final String id;
  final int periodYear;
  final int periodMonth;
  final double grossAmount;
  final double netAmount;
  final double incomeTax;
  final String status;

  const Payslip({
    required this.id,
    required this.periodYear,
    required this.periodMonth,
    required this.grossAmount,
    required this.netAmount,
    required this.incomeTax,
    required this.status,
  });

  String get period =>
      '$periodYear-${periodMonth.toString().padLeft(2, '0')}';

  factory Payslip.fromJson(Map<String, dynamic> j) => Payslip(
        id: j['id'] as String,
        periodYear: j['periodYear'] as int? ?? 0,
        periodMonth: j['periodMonth'] as int? ?? 0,
        grossAmount: (j['grossAmount'] as num? ?? 0).toDouble(),
        netAmount: (j['netAmount'] as num? ?? 0).toDouble(),
        incomeTax: (j['incomeTax'] as num? ?? 0).toDouble(),
        status: j['status'] as String? ?? '',
      );
}
